src/wallet/asyncrpcoperation_shieldcoinbase.cpp

```diff
@@ -157,303 +107,136 @@ void AsyncRPCOperation_shieldcoinbase::main() {
-
-    LogPrint("zrpcunsafe", "%s: creating joinsplit at index %d (vpub_old=%s, vpub_new=%s, in[0]=%s, in[1]=%s, out[0]=%s, out[1]=%s)\n",
-            getId(),
-            tx_.vJoinSplit.size(),
-            FormatMoney(info.vpub_old), FormatMoney(info.vpub_new),
-            FormatMoney(info.vjsin[0].note.value()), FormatMoney(info.vjsin[1].note.value()),
-            FormatMoney(info.vjsout[0].value), FormatMoney(info.vjsout[1].value)
-            );
-
-    // Generate the proof, this can take over a minute.
-    std::array<libzcash::JSInput, ZC_NUM_JS_INPUTS> inputs
-            {info.vjsin[0], info.vjsin[1]};
-    std::array<libzcash::JSOutput, ZC_NUM_JS_OUTPUTS> outputs
-            {info.vjsout[0], info.vjsout[1]};
-    std::array<size_t, ZC_NUM_JS_INPUTS> inputMap;
-    std::array<size_t, ZC_NUM_JS_OUTPUTS> outputMap;
-
-    uint256 esk; // payment disclosure - secret
-
-    assert(mtx.fOverwintered && (mtx.nVersion >= SAPLING_TX_VERSION));
-    JSDescription jsdesc = JSDescriptionInfo(
-            joinSplitPubKey_,
-            anchor,
-            inputs,
-            outputs,
-            info.vpub_old,
-            info.vpub_new
-    ).BuildRandomized(
-            inputMap,
-            outputMap,
-            !this->testmode,
-            &esk); // parameter expects pointer to esk, so pass in address
-    {
-        auto verifier = ProofVerifier::Strict();
-        if (!(verifier.VerifySprout(jsdesc, joinSplitPubKey_))) {
-            throw std::runtime_error("error verifying joinsplit");
+        if (maxedOutFlag) {
+            remainingValue += nValue;
         }
     }
 
-    mtx.vJoinSplit.push_back(jsdesc);
+    spendable.LimitTransparentUtxos(numUtxos);
 
-    // Empty output script.
-    CScript scriptCode;
-    CTransaction signTx(mtx);
-    std::vector<CTxOut> allPrevOutputs;
-    for (ShieldCoinbaseUTXO & t : inputs_) {
-        allPrevOutputs.emplace_back(t.amount, t.scriptPubKey);
-    }
-    PrecomputedTransactionData txdata(signTx, allPrevOutputs);
-    uint256 dataToBeSigned = SignatureHash(scriptCode, signTx, NOT_AN_INPUT, SIGHASH_ALL, 0, consensusBranchId, txdata);
-
-    // Add the signature
-    ed25519::sign(
-        joinSplitPrivKey_,
-        {dataToBeSigned.begin(), 32},
-        mtx.joinSplitSig);
-
-    // Sanity check
-    if (!ed25519::verify(
-        mtx.joinSplitPubKey,
-        mtx.joinSplitSig,
-        {dataToBeSigned.begin(), 32}))
-    {
-        throw std::runtime_error("ed25519_verify failed");
+    // TODO: Don’t check this outside of `PrepareTransaction`
+    if (shieldingValue < fee_) {
+        ThrowInputSelectionError(
+                InvalidFundsError(shieldingValue, InsufficientFundsError(fee_)),
+                ztxoSelector_,
+                strategy_);
     }
 
-    CTransaction rawTx(mtx);
-    tx_ = rawTx;
+    // FIXME: This should be internal, and shouldn’t be a payment, but more like a change address,
+    //        as we don’t know the amount at this point.
```

Comment:
Thinking through this, nothing blocking; maybe improve the comment if making other changes:
* we have limited transparent UTXOs at this point, so we should be able to accurately compute the ZIP 317 fee here; I don't fully understand the "we don't know the amount at this point" comment.
* It would be possible to now make the recipient address of a `z_shieldcoinbase` transaction optional at the RPC level, using the change address for the account if not provided. However, this would slightly alter change pool determination, because we'd need to special-case the situation where *no* output pools are specified. For now it makes sense to me to just keep this as a payment. It's not critical here, but maybe we should add something to the wallet tx builder that allows the caller to force the change amount to be zero, ensuring that the inputs are fully consumed?
* I don't fully understand how the P2SH case factors in here (and elsewhere in ZIP 317 fee computation); for multisig P2SH the wallet would have to control all of the keys needed to make the signatures, for other P2SH scripts I guess it's possible that the wallet could store nonstandard spend scripts, and the low-level transaction builder can then spend those UTXOs correctly. My suspicion is that P2SH spends will usually underpay ZIP 317 fees for now, but that's okay because they'll still be close enough to be prioritized highly.

Reply:
Ah, this is a good point.

My commit that’s blocked by this one introduces a `NetAmountPayment`, which generalizes `MergeToAddressRecipient` and does the fee & send_amount calculation in WalletTxBuilder for the remaining operations. But _maybe_ we can make a smaller change for now that calculates the fee for these outside WalletTxBuilder.

I think the thing we don’t know, when shielding to a UA, is which pool it’s going to. But … it can’t be transparent or Sprout, and both Sapling and Orchard pad to 2 outputs, so that’s enough info to calculate the fee. For mergetoaddress, it’s maybe a bit more complicated, as there’s also the chance that a UA recipient ends up going to a transparent receiver (e.g., merging Sprout to an Orchard+transparent UA), which would have a lower fee. But the existing changes I’m making for this already break `ResolvePayment` out of `ResolveInputsAndPayments`, so if I make that public, then mergetoaddress can also calculate its own fee correctly.

Anyway, I’ll have to consider it compared to my about to be pushed PR, so I think I’ll keep it as a separate PR, so the approach can be evaluated.

Reply:
The idea is that the `fee_` we've computed at this point won't have been able to take into account the actual size of the spend script that the low-level transaction builder ultimately uses to spend from the P2SH input. So we'll compute a fee that _may_ underestimate the ultimate transaction size, and consequently the ZIP 317 conventional fee for that tx.

Reply:
Ah, building `CTxIn`/`Out` is pretty trivial. So, WalletTxBuilder currently does that just to calculate the fee from the serialization and throw it away. So, I think we’d get that right no matter where we do it (as long as we have something like UTXOs and not just a count).